Fix the NPD occurred in short-circuit evaluation.

QiAnXinCodeSafe opened the pull request:
When `encname` is a null pointer, there will be a null pointer dereference in the short-circuit evaluation.

Found by the static analyzer of Qihoo360 CodeSafe Team!

QiAnXinCodeSafe committed:
Fix the NPD occurred in short-circuit evaluation.

When `encname` is a null pointer, there will be a null pointer dereference in the short-circuit evaluation.

Found by the static analyzer of Qihoo360 CodeSafe Team!

HughWilliams commented:
Thanks for you proposed patch and other reports of other possible NPD occurrences, which are being reviewed by development, who will report back on these individually ...

QiAnXinCodeSafe committed:
Fix the non-fatal issue about the operand of the `sizeof`.

Result of `s_realloc`, i.e. `realloc` is converted to `char**`, which is incompatiable with the sizeof operand type `char**`.

Given the code below, the operand of sizeof is `int`, the allocated memory type will be `int*`. In the same way, if the operand of sizeof is `char**`, the allocated memory type will be `char***`.
```
int *ptr = (int*)malloc(sizeof(int));
```

The operand of sizeof, whether `char*` or `char**`, result in the same program behavior. But use the consistent manner will improve the readability.

pkleef commented:
Thanks for reporting this issue to us.
